posts/2021-01-06-ocp-log-forwarding.adoc

```diff
@@ -0,0 +1,425 @@
+* The `pipelines` section of the file specifies which log types are sent to the configured outputs.
++
+The following sample `log-forward-instance.yaml` file defines two `outputs`: a secure connection to a Fluentd server named `fluentd-server-secure` and an insecure connection to another Fluentd server named `fluentd-server-insecure`. Note the different protocols that are defined for the _url_ in each output (tls vs tcp). Remove the output type that you do not want to use in your file. Each log type and its corresponding output reference is defined in the `pipelines` section. The `default` output reference points to the internal OpenShift Container Platform Elasticsearch instance:
++
+```
+apiVersion: logging.openshift.io/v1
+kind: ClusterLogForwarder
+metadata:
+  name: instance
+  namespace: openshift-logging
+spec:
+  disableDefaultForwarding: true
+  outputs:
+   - name: fluentd-server-secure
+     type: fluentdForward
+     url: 'tls://fluntdserver.security.example.com:24224' # Substitute with FQDN or IP address
+     secret:
+        name: secure-forward
+   - name: fluentd-server-insecure
+     type: fluentdForward
+     url: 'tcp://fluntdserver.home.example.com:24224' # Substitute with FQDN or IP address
+  pipelines:
+   - name: container-logs
+     inputRefs:
+     - application
+     outputRefs:
+     - default
+     - fluentd-server-secure
+   - name: infra-logs
+     inputRefs:
+     - infrastructure
+     outputRefs:
+     - default
+   - name: audit-logs
+     inputRefs:
+     - audit
+     outputRefs:
+     - default
+
+```
++
+For more information, see link:https://docs.openshift.com/container-platform/4.6/logging/cluster-logging-external.html#cluster-logging-collector-log-forward-fluentd_cluster-logging-external[Forwarding logs using the Fluentd forward protocol]. Make sure to consult the documentation for your version of OpenShift, as the pipeline configuration for version 4.6 is slightly different from the preceding versions.
+
+
+[start=2]
+. Create the Log Forwarding instance inside your OpenShift cluster:
++
+[source]
+----
+[root@ocp ~]# oc create -f log-forward-instance.yaml
+----
++
+
+. **OpenShift 4.3 to 4.5 only:** Annotate the ClusterLogging instance to enable the Log Forwarding API.
++
+[source]
+----
+[root@ocp ~]# oc annotate clusterlogging -n openshift-logging instance clusterlogging.openshift.io/logforwardingtechpreview=enabled
+----
++
+
+[start=3]
+. To check whether the logs are forwarded to the specified outputs, run the following command:
++
+[source]
+----
+[root@ocp ~]# oc -n openshift-logging get cm fluentd -o json | jq -r '.data."fluent.conf"' > fluentd-with-logfowarding.conf
+----
++
+This command displays the ConfigMap configuration for Fluentd inside OpenShift. Check whether the outputs are defined inside the configuration file.
+
+* In the following configuration file example, you can see the `elasticsearch` and `fluent-forward` outputs that are referenced by `@label`:
```

Comment:
The intro sentence here still makes a reference to the `elasticsearch` output, which is only valid for 4.3-.46, and has been removed from the examples. Suggest the following edit:

In the following ConfigMap file example, the `FLUENTD_FORWARD`  output type is specified for the `CONTAINER_LOGS` pipeline:

I don't think we need to mention DEFAULT here, as it's not the focus. However, want to check-  In our yaml example, it says 

`disableDefaultForwarding: true` 

is the default output still available in that case? Probably ok but just want to make sure the examples mixing the default and custom outputs are viable.

Reply:
A good catch on the `disableDefaultForwarding: true` value.
This is a relic from the 4.3-4.5 configurations. It would've disabled the _default_ behavior and that's why I assume the need to define Elasticsearch output in those versions. However it would seem that  attribute is not recognized anymore in 4.6 and nothing _happened_ when I was running through.